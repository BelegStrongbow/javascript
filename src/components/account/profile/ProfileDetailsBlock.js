import React from "react";
import PropTypes from "prop-types";
import { injectIntl, intlShape, FormattedMessage, defineMessages } from "react-intl";
import styled from "styled-components";

import defaults from "../../../config/defaults.json";
import colors from "yoast-components/style-guide/colors.json";

// Components.
import UploadUserImage from "./UploadUserImage";
import { InputField } from "../../InputField";
import { StyledLabel } from "../../Labels";

const FormGroup = styled.form`
	display: flex;
	flex-wrap: wrap;
	width: 100%;
	justify-content: space-between;
	text-align: left;
`;

const LabelBlock = styled.div`
	width: 100%;
`;

const NameBlock = styled( LabelBlock )`
	width: 70%;
	padding-left: 20px;

	@media screen and ( max-width: ${ defaults.css.breakpoint.mobile }px ) {
		width: 100%;
	}

	div:last-of-type{
		margin-bottom: 0;
	}
`;

const AvatarBlock = styled.div`
	width: 30%;
	margin: auto;
	margin-top: 25px;
	text-align: center;
`;

const TextInput = styled( InputField )`
	background-color: ${ colors.$color_grey_light };
`;

const messages = defineMessages( {
	labelFirstName: {
		id: "profile.label.firstName",
		defaultMessage: "First name",
	},
	labelLastName: {
		id: "profile.label.lastName",
		defaultMessage: "Last name",
	},
} );

/**
 * A component for entering the first- and last name and
 * uploading a profile image.
 */
class ProfileDetailsBlock extends React.Component {

	constructor( props ) {
		super( props );

		this.state = {
			userFirstName: this.props.userFirstName,
			userLastName: this.props.userLastName,
			imageFile: null,
		};

		this.onUpdateFirstName = this.onUpdateField.bind( this, "userFirstName" );
		this.onUpdateLastName = this.onUpdateField.bind( this, "userLastName" );
		this.onFileUpload = this.onFileUpload.bind( this );
		this.handleSubmit = this.handleSubmit.bind( this );
	}

	/**
	 * Updates the given field in the state with the target value in the
	 * given event.
	 * @param {string} field the field that needs to be updated.
	 * @param {object} event the event from which to get the new value.
	 * @returns {void}
	 */
	onUpdateField( field, event ) {
		this.setState( {
			[ field ]: event.target.value,
		} );
	}

	/**
	 * Calls the onSubmit function given in the props
	 * with the entered first name, last name and image file.
	 * @returns {void}
	 */
	handleSubmit() {
		this.props.onSubmit( this.state.userFirstName, this.state.userLastName, this.state.imageFile );
	}

	/**
	 * Updates the state with the uploaded image file.
	 * @param {File} file the uploaded file.
	 * @returns {void}
	 */
	onFileUpload( file ) {
		this.setState( {
			imageFile: file,
		} );
	}

	render() {
		return <FormGroup onSubmit={ this.props.onSubmit }>
			<AvatarBlock>
				<UploadUserImage onFileUpload={ this.onFileUpload } />
			</AvatarBlock>

			<NameBlock id="left">
				<StyledLabel htmlFor="first-name">
					<FormattedMessage
						id={ messages.labelFirstName.id }
						defaultMessage={ messages.labelFirstName.defaultMessage }
					/>
				</StyledLabel>
				<TextInput
					width="100%"
					margin-right="5px"
					id="first-name"
					name="first name"
					type="text"
					value={ this.state.userFirstName }
					onChange={ this.onUpdateFirstName }
				/>
				<StyledLabel htmlFor="last-name">
					<FormattedMessage
						id={ messages.labelLastName.id }
						defaultMessage={ messages.labelLastName.defaultMessage }
					/>
				</StyledLabel>
				<TextInput
					width="100%"
					id="last-name"
					name="last name"
					type="text"
					value={ this.state.userLastName }
					onChange={ this.onUpdateLastName }
				/>
			</NameBlock>
			{ this.props.children }
		</FormGroup>;
	}
}

export default injectIntl( ProfileDetailsBlock );

ProfileDetailsBlock.propTypes = {
	intl: intlShape.isRequired,
	children: PropTypes.any,
	onSubmit: PropTypes.func,
	onUploadAvatar: PropTypes.func,
	userFirstName: PropTypes.string,
	userLastName: PropTypes.string,
};

ProfileDetailsBlock.defaultProps = {
	userFirstName: "",
	userLastName: "",
};
