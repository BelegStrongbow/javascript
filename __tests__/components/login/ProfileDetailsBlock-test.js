import React from 'react';
import ProfileDetailsBlock from "../../../src/components/account/profile/ProfileDetailsBlock";
import { createComponentWithIntl } from "../../../utils";

test( 'The ProfileDetailsBlock component matches the snapshot', () => {
	const component = createComponentWithIntl(
		<ProfileDetailsBlock />
	);

	let tree = component.toJSON();
	expect( tree ).toMatchSnapshot();
} );