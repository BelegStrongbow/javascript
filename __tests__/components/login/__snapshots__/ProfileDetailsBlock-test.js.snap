// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ProfileDetailsBlock component matches the snapshot 1`] = `
.c3 {
  -webkit-flex: 0 0 125px;
  -ms-flex: 0 0 125px;
  flex: 0 0 125px;
  height: 125px;
  width: 125px;
  border-radius: 50%;
  margin-right: 12px;
  object-fit: cover;
}

.c2 {
  display: inline-block;
  position: relative;
  width: 125px;
  height: 125px;
}

.c4 {
  position: absolute;
  bottom: 0;
  left: 0;
  width: 100%;
  height: 100%;
  border-radius: 50%;
  background: linear-gradient(rgba(0,0,0,0) 75%,rgba(0,0,0,0.5) 75%);
}

.c5 {
  position: absolute;
  bottom: 0;
  left: 0;
  width: 100%;
  padding: 8px;
  background: none;
  border: none;
  color: white;
  font-size: 12px;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  text-align: center;
}

.c5:hover,
.c5:focus {
  color: white;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c6 {
  margin: 0;
  width: 100%;
  text-align: center;
  font-size: 12px;
  color: black;
}

.c11 {
  box-shadow: inset 0 1px 8px 0 rgba(0,0,0,0.3);
  background: #fff;
  border: 0;
  padding: 0 0 0 10px;
  font-size: 1em;
  font-family: "Open Sans",sans-serif;
  height: 3em;
  width: 100%;
}

.c9 {
  display: inline-block;
  font-size: 1em;
  margin: 16px 0 8px;
  font-weight: 700;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  width: 100%;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  text-align: left;
}

.c8 {
  width: 100%;
}

.c7 {
  width: 70%;
  padding-left: 20px;
}

.c7 div:last-of-type {
  margin-bottom: 0;
}

.c1 {
  width: 30%;
  margin: auto;
  margin-top: 25px;
  text-align: center;
}

.c10 {
  background-color: #f1f1f1;
}

@media screen and ( max-width:800px ) {
  .c7 {
    width: 100%;
  }
}

<form
  className="c0"
  onSubmit={undefined}
>
  <div
    className="c1"
  >
    <div
      className="c2"
      size="125px"
    >
      <img
        alt="Default profile image"
        className="c3"
        onError={[Function]}
        size="125px"
        src="https://s3.amazonaws.com/yoast-my-yoast/default-avatar.png"
      />
      <span
        className="c4"
      >
        <button
          aria-label="Change profile image"
          className="c5"
          onClick={[Function]}
          type="button"
        >
          <span>
            Change
          </span>
        </button>
      </span>
      <p
        className="c6"
      >
        <span>
          Maximum file size is 5 megabytes
        </span>
      </p>
      <input
        accept="image/png, image/jpeg, image/gif"
        aria-hidden={true}
        onChange={[Function]}
        style={
          Object {
            "display": "none",
          }
        }
        type="file"
      />
    </div>
  </div>
  <div
    className="c7 c8"
    id="left"
  >
    <label
      className="c9"
      htmlFor="first-name"
    >
      <span>
        First name
      </span>
    </label>
    <input
      className="c10 c11"
      id="first-name"
      name="first name"
      onChange={[Function]}
      type="text"
      value=""
      width="100%"
    />
    <label
      className="c9"
      htmlFor="last-name"
    >
      <span>
        Last name
      </span>
    </label>
    <input
      className="c10 c11"
      id="last-name"
      name="last name"
      onChange={[Function]}
      type="text"
      value=""
      width="100%"
    />
  </div>
</form>
`;
